### PR TITLE
Update RxJS to v5.0.0

### PR DESCRIPTION
**Pull request by blesh**
### EDIT: updated 12/02/17

This just updates this library to use RxJS 5. Primary motivation for this being debugging and performance.

NOTE: The use of Rx's `expand` had to have the `Rx.Scheduler.queue` scheduler passed to it explicitly, to insure the trampolining behavior that falcor-router was relying on. This was the default with Rx 4 and under. Overall, this is probably a good thing, as falcor router could theoretically be asked to crawl sufficiently large enough graphs to cause a stack overflow in some cases.

This just updates this library to use RxJS 5. Primary motivation for this being debugging and performance.

(edit, edit): no longer need Symbol.observable patching.

### CODE REVIEW

**Comment by jhamlet** [comment]
@blesh what's the work-a-round for getting the `ModelResponse` to work with RxJs@5+?

**Comment by blesh** [comment]
@jhamlet you might be able to patch ModelResponse to have a proper implementation of Symbol.observable. (you can find one at https://github.com/blesh/symbol-observable) Otherwise, I'd wait for that PR to land.

File: test/unit/functional/collapse-and-batch.spec.js [hunk 1]
```diff
@@ -201,21 +201,21 @@ describe('Collapse and Batch', function() {
 
     // HACK: This test is not passing
     // NOTE: I made the test easier to read by simplifying it. I'm still not sure what it's testing.
-    it('should validate batching/collapsing makes two request since its onNextd without toArray().', function(done) {
+    it.only('should validate batching/collapsing makes two request since its onNextd without toArray().', function(done) {
```
Comment (hgwood) [inline test/unit/functional/collapse-and-batch.spec.js]: Is this on purpose ?

File: test/unit/functional/collapse-and-batch.spec.js [hunk 2]
```diff
@@ -199,25 +199,23 @@ describe('Collapse and Batch', function() {
...
-                                path: ['lists', id],
-                                value: $ref('two.be[956]')
-                            };
-                        }
-                        return {
-                            path: ['lists', id],
-                            value: $ref('lists[0]')
-                        };
-                    });
+                return Observable.of(
```
Comment (trxcllnt) [inline test/unit/functional/collapse-and-batch.spec.js]: @blesh this change is wrong -- if `lists[1]` points to `lists[0]`, this get handler should be re-entered for `list[0]`. If it returns two values for `list[0]` (and the router is working correctly), this route should infinitely recurse. The current behavior only returns values for indexes specified in the `ids` position. The new test should still do this.

File: test/unit/functional/collapse-and-batch.spec.js [hunk 3]
```diff
@@ -238,7 +236,7 @@ describe('Collapse and Batch', function() {
             get([['lists', [0, 1], 'summary']]);
         var count = 0;
         obs.
-            doAction(function(res) {
+            subscribe(function(res) {
```
Comment (trxcllnt) [inline test/unit/functional/collapse-and-batch.spec.js]: the pattern of `doAction(() => expect(...)).subscribe(noop, done, done)` ensures errors thrown by `expect()` get forwarded to the `done` handler. If expectations happen in the subscribe handlers, the `done` handler never gets invoked.

**Comment by trxcllnt** [comment]
@blesh tbh expand scheduling on the queue scheduler only matters for routes which return data synchronously (values, or sync Observables). We're running the router in production on a massive domain graph, most of which is backed by synchronous caches, and haven't had any issues. The moment an async route is encountered (which usually happens at least once), the stack unwinds for any sync expands. In our experience, the queue scheduling is prohibitively expensive for running falcor router on the client, and produces nasty flame graphs for our servers. Perhaps the asap scheduler is more appropriate (for the client), though still produces difficult flame graphs.

**Comment by blesh** [comment]
> the queue scheduling is prohibitively expensive

@trxcllnt are you saying the queue itself gets expensive?

File: test/integration/get.spec.js [hunk 4]
```diff
@@ -1,54 +1,47 @@
...
+                          summary: 75
+                      }
+                  }
+              });
+            }, done, function() {
                 expect(called).to.be.ok;
-            }).
-            subscribe(noOp, done, done);
+                done();
+            });
```
Comment (jhusain) [inline test/integration/get.spec.js]: I don't think done will get invoked if the assertion in next throws. Errors caught in a subscribe are not referred to error. The `doAction` method ensured this before.

File: test/integration/get.spec.js [hunk 5]
```diff
@@ -1,54 +1,47 @@
...
-        Rx.Observable.
-            of(model.get('genreLists[0].summary')).
-            flatMap(function(obs) {
-                return obs;
-            }).
-            doAction(function(x) {
+        model.get('genreLists[0].summary').
+            subscribe(function (x) {
                 called = true;
                 expect(x).to.deep.equals({
```
Comment (jhusain) [inline test/integration/get.spec.js]: I don't think done will get invoked if the assertion in next throws. Errors caught in a subscribe are not referred to error. The doAction method ensured this before.

File: test/integration/get.spec.js (hunk 4 above)
Comment (blesh) [inline test/integration/get.spec.js]: Yup. I think you're right, I'll change it back.

File: test/unit/functional/collapse-and-batch.spec.js (hunk 2 above)
Comment (blesh) [inline test/unit/functional/collapse-and-batch.spec.js]: Yeah, @jhusain and I had gone over this and confirmed that it was never going to be any different, but I see what you're saying. I'll change it back.

**Comment by blesh** [comment]
FYI: @jhusain @trxcllnt I've updated this PR to reflect your requested changes

**Comment by jhusain** [comment]
LGTM

**Pull request merged**